File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.py
def solution(id_list, report, k):
    n = len(id_list)
    answer = [0] * n
    
    id_dict = dict()
    for i in range(n):
        id_dict[id_list[i]] = i
    
    # 신고 현황 2차원 배열에 저장
    result = [[0] * n for _ in range(n)]
    for r in report:
        a, b = r.split()
        a_idx, b_idx = id_dict[a], id_dict[b]
        result[a_idx][b_idx] = 1
    
    # 신고당한 횟수 파악
    reported_time = [0] * n
    for j in range(n):
        for i in range(n):
            if result[i][j] != 0:
                reported_time[j] += 1
    
    # k번 이상 신고 당한 유저 찾고 -> 그 유저를 신고한 유저 찾기
    for j in range(n):
        if reported_time[j] >= k:   # 신고 횟수 k번 이상인 경우
            for i in range(n):
                if result[i][j] != 0:
                    answer[i] += 1
    return answer

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.java
import java.util.*;

class Solution {
    public int[] solution(String[] id_list, String[] report, int k) {
        int[] answer = new int[id_list.length];
        
        HashMap<String, Integer> idMap = new HashMap<>();
        for (int i = 0; i < id_list.length; i++) {
            idMap.put(id_list[i], i);
        }
        
        HashMap<String, Set<String>> reportMap = new HashMap<>();
        for (String rep : report) {
            String[] input = rep.split(" ");
            
            if (!reportMap.containsKey(input[1])) {
                Set<String> set = new HashSet<>();
                reportMap.put(input[1], set);  
            } 
            
            reportMap.get(input[1]).add(input[0]);
        }
        
        for (Map.Entry<String, Set<String>> entry : reportMap.entrySet()) {
            if (entry.getValue().size() >= k) {
                // 신고한 유저들 목록
                String[] names = entry.getValue().toArray(new String[0]);
                for (String name : names) {
                    answer[idMap.get(name)]++;
                }
            }
        }
        
        return answer;
    }
}